src/Workspaces/Core/Portable/Workspace/Solution/Solution.cs

```diff
@@ -259,22 +245,17 @@ private static Project CreateProject(ProjectId projectId, Solution solution)
         {
             if (syntaxTree != null)
             {
-                // is this tree known to be associated with a document?
-                var docId = DocumentState.GetDocumentIdForTree(syntaxTree);
-                if (docId != null && (projectId == null || docId.ProjectId == projectId))
+                var documentState = State.GetDocumentState(syntaxTree, projectId);
+
+                if (documentState is SourceGeneratedDocumentState)
                 {
-                    // does this solution even have the document?
-                    // We call TryGetSourceGeneratedDocumentForAlreadyGeneratedId if it's not a regular document;
-                    // we know that if we already have the tree that means generation completed, so that's safe to call.
-                    var document = this.GetDocument(docId) ?? this.GetProject(docId.ProjectId)?.TryGetSourceGeneratedDocumentForAlreadyGeneratedId(docId);
-                    if (document != null)
-                    {
-                        // does this document really have the syntax tree?
-                        if (document.TryGetSyntaxTree(out var documentTree) && documentTree == syntaxTree)
-                        {
-                            return document;
-                        }
-                    }
+                    var generatedDocument = this.GetRequiredProject(documentState.Id.ProjectId).TryGetSourceGeneratedDocumentForAlreadyGeneratedId(documentState.Id);
```

Comment:
How is this different from the call that `State.GetDocumentState` already made `TryGetSourceGeneratedDocumentForAlreadyGeneratedId`?

Reply:
Ah, yes, this needs clarification. The TryGetSourceGeneratedDocumentForAlreadyGeneratedId in GetDocumentState is fetching the *state* and not the wrapper object, this call here is fetching the wrapper object. Since the Project object needs to be the one creating the wrappers for object identity reasons, we have to defer back here.

I'll add a comment here clarifying, and will also rename the TryGetSourceGeneratedDocumentForAlreadyGeneratedId that fetches the state to at least help make it clear they're not the entire same thing.

Reply:
@davidwengier: PR updated, let me know if the comment clarifies the understanding.

Reply:
The new name is much clearer, thanks